operator/v1alpha1/types_externaldns.go

```diff
@@ -0,0 +1,366 @@
+	//
+	// +optional
+	AnnotationFilter *string `json:"annotationFilter,omitempty"`
+
+	// Namespace instructs externalDNS to only acknowledge
+	// source resource instances in a specific namespace.
+	//
+	// +optional
+	Namespace *corev1.LocalObjectReference `json:"namespace,omitempty"`
+
+	// HostnameAnnotationPolicy specifies whether or not externalDNS
+	// should ignore the "external-dns.alpha.kubernetes.io/hostname"
+	// annotation, which overrides DNS hostnames on a given source resource.
+	//
+	// The following values are accepted:
+	//
+	//  "Ignore": Ignore any hostname annotation overrides.
+	//  "Allow": Allow all hostname annotation overrides.
+	//
+	// The default behavior of the externalDNS is "Ignore".
+	//
+	// +optional
+	HostnameAnnotationPolicy *HostnameAnnotationPolicy `json:"hostnameAnnotation,omitempty"`
+
+	// Service describes source configuration options specific
+	// to the service source resource.
+	//
+	// +optional
+	Service *ExternalDNSServiceSourceOptions `json:"service,omitempty"`
+
+	// Service describes source configuration options specific
+	// to the CRD source resource>
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=ignore;allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the externalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// PublishPolicy determines what kinds of Service
+	// resources are published to the given DNS provider.
+	//
+	// The following publishing policies are available:
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service resources that externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of externalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what type of Service resources
+	// are watched by externalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// +optional
+	ServiceType *corev1.ServiceType `json:"serviceType,omitempty"`
```

Comment:
Actually, this can be a list! Surprisingly, ExternalDNS does accept a list for this command line argument.
https://github.com/kubernetes-sigs/external-dns/blob/master/source/service.go#L67
Fixed! Thanks.

Reply:
Actually, looks like the verify job yells at me for using `// +kubebuilder:validation:UniqueItems=true` here.

The CustomResourceDefinition "externaldns.operator.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[source].properties[service].properties[serviceType].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic

For now ill leave the list change but drop the UniqueItems tag. Ill bring this up during API review.